src/Keboola/S3Extractor/Finder.php

```diff
@@ -55,29 +58,116 @@ public function __construct(Config $config, State $state, LoggerInterface $logge
+        // Read sorted metadata
+        $sortedFile = fopen($sortedFilePath, "r");
+        if (!$sortedFile) {
+            throw new \RuntimeException(sprintf('Cannot open sorted file "%s".', $sortedFilePath));
+        }
+        try {
+            $i = 0;
+            while (($line = fgets($sortedFile, 20480)) !== false) {
+                // Skip the first word "<timestamp><key>", separated by NUL char
+                strtok($line, "\0");
+                $serialized = (string)strtok("");
+
+                /** @var File $file */
+                $file = unserialize(base64_decode($serialized));
+
+                yield $file;
+                if ($this->limit > 0 && ++$i >= $this->limit) {
+                    break;
+                }
+            }
+        } finally {
+            fclose($sortedFile);
+        }
+    }
+
+    /**
+     * sortLines by the first word: "<timestamp><key>"
+     */
+    private function sortLines(string $tmpFilePath): string
+    {
+        $sortedFilePath = $tmpFilePath . ".sorted";
+        $args = [
+            "sort", // the sort command has small memory requirements, it uses temp files for sorting, instead of memory
+            "--stable",
+            "--output", $sortedFilePath,
+            "-t", '\0', // words are separated by NUL character
+            "-k", "1,1", // sort lines by the first word "<timestamp><key>" (start=1, end=1)
+            "--parallel=1", // optimizes memory usage
+            $tmpFilePath,
+        ];
+        $env = ["LC_ALL" => "C"];
+        (new Process($args, null, $env))->mustRun();
+        return $sortedFilePath;
+    }
+
+    /**
+     * @param \Iterator|File[] $files
+     */
+    private function dumpFilesListToTmpFile(\Iterator $files): string
+    {
+        // Write all metadata from the generator to a temporary file, to prevent memory issues.
+        $tmpFilePath = tempnam(sys_get_temp_dir(), 'aws-files');
+        if (!$tmpFilePath) {
+            throw new \RuntimeException("Cannot create a temp file.");
+        }
+
+        $tmpFile = fopen($tmpFilePath, "w");
+        if (!$tmpFile) {
+            throw new \RuntimeException(sprintf('Cannot open temp file "%s".', $tmpFilePath));
+        }
 
-        $filesToDownload = iterator_to_array($this->listFiles());
-        $this->logger->info(sprintf(
-            'Found %s file(s)',
-            count($filesToDownload)
-        ));
+        try {
+            foreach ($files as $file) {
+                // Write each file metadata as: <timestamp><key>\0<serialized S3File object>\n
+                // The base64 encoding is used to prevent new lines in the serialized object.
+                fwrite($tmpFile, str_pad((string)$file->getTimestamp(), self::TIMESTAMP_STR_LENGTH, "0"));
+                fwrite($tmpFile, $file->getKey());
+                fwrite($tmpFile, "\0");
+                fwrite($tmpFile, base64_encode(serialize($file)));
+                fwrite($tmpFile, "\n");
```

Comment:
V subore sa zoradzuju riadky, oddelovac `\n`.

Zoradzuje sa podla prveho slova `<timestamp><key>`, tato logika zodpoveda povodnemu kodu:
https://github.com/keboola/aws-s3-extractor/blob/10d50d4eee66af4076da59705f331ae312cef11d/src/Keboola/S3Extractor/Extractor.php#L254-L260

Reply:
`timestamp` ma pevnu dlzku, ak by sa nejak dal nastavit vlastny, a niekto by tam dal napr. `123`, tak to bude `000000000123`.